File: crackingthecodeinterview/trees/minimumtree.cpp
#include <iostream>
#include <vector>
#include "node.h"

using namespace std;

template <typename T>
struct MinTree{
    
        Node<T>* root;

        MinTree(void) : root(NULL) {}
        
        Node<T>* BuildMinTree(T*, int, int);
        void     Traverse(Node<T>*, ostream&);

        friend ostream &operator<<(ostream& os, MinTree& mt)
        {
            mt.Traverse(mt.root,os);
            return os;
        } 
};

template <typename T>
Node<T>*
MinTree<T>::BuildMinTree(T* in, int start, int end)
{
    if(start > end || end == 0)
        return NULL;

    int middle    = ((start+end)/2);
    Node<T>* root = new Node<T>(in[middle]);
    root->left    = BuildMinTree(in,start,middle-1); 
    root->right   = BuildMinTree(in,middle+1,end);
    return root;
}

template <typename T>
void
MinTree<T>::Traverse(Node<T>* in, ostream& os)
{
    os << "->" << *in;
    cerr << in->value << endl;
    Traverse(in->left,os);
    Traverse(in->right,os);
}

int
main()
{
    MinTree<int> tree;
    int array[6] = {3,4,5,6,7,8};
    tree.BuildMinTree(array,0,5);
    
    cout << tree << endl;

    return 1;
}

File: crackingthecodeinterview/trees/minimumtree.py
class Node:

    def __init__(self,data=None):
        self.left   = None
        self.right  = None
        self.parent = None
        self.data   = data 
        self.depth  = 0

    def __str__(self):
        return str(data)

    def __eq__(self,other):
        if is_instance(other, type(self)):
            return (other.left == self.left) and (other.right == self.right) and (other.parent == self.parent) and (self.data == other.data) and (self.depth == other.depth)

    def __lt__(self):
        if is_instance(other,type(self)):
            return self.data < other.data

    def __gt__(self):
        if is_instance(other,type(self)):
            return self.data > other.data

class MinTree:
    
    def __init__(self):
        self.root = None; 

    def BuildMinTree(self, value, start, end):
        if start > end or end == 0:
            return
    
        middle    = (start+end)/2
        middleval = value[int(middle)]

        root = Node(middleval)
        root.left = self.BuildMinTree(value,start,middle-1)
        root.right = self.BuildMinTree(value,middle+1,end)
        return root
         

def main():

    tree = MinTree()
    inar = [3,4,5,6,7,8]
        
    tree.root = tree.BuildMinTree(inar,0,5)
    
    tempnode = tree.root

main()
